### PR TITLE
feat: Add ReassignmentHandler which is notified on client reassignment

### LINKED ISSUE #869: Provide support for client partition reassignment handlers
We should support client partition reassignment handlers to allow clients to react to lost assignments. 

It would be a nice to have if messages on partitions being assigned away could call nack() without needing to install a custom handler.

### PR DESCRIPTION
**Pull request by dpcollins-google**
This enables clients to respond to a reassignment by cancelling outstanding actions and nacking all messages.

Fixes #869

### CODE REVIEW

**Review by tmdiep — APPROVED** [review]

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/ReassignmentHandler.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+package com.google.cloud.pubsublite.cloudpubsub;
+
+import com.google.cloud.pubsublite.Partition;
+import java.util.Set;
+
+/**
+ * A ReassignmentHandler is called any time a new partition assignment is received from the server.
+ * It will be called with both the previous and new assignments as decided by the backend.
```
[the remark this answers is not in the document]
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/ReassignmentHandler.java:24]: This CL specifically adds fencing for nacks and tests for this behavior. The test case added to SinglePartitionSubscriberTest.java confirms that calling stopAsync is sufficient to fence nacks from having any effect. This is a guarantee that nacks are allowed after shutdown- acks may or may not be processed. The wording specifies "will be" which is equivalent to a guarantee.

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/ReassignmentHandler.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * <p>The client library will not acknowledge the assignment until handleReassignment returns. The
+ * assigning backend will not assign any of the partitions in `before` to another server unless the
+ * assignment is acknowledged, or a client takes too long to acknowledged (currently 30 seconds from
+ * the time the assignment is sent from server's point of view).
+ *
+ * <p>Because of the above, as long as reassignment handling is processed quickly, it can be used to
+ * abort outstanding operations on partitions which are being assigned away from this client, or to
+ * pre-warm state which will be used by the subscriber handler.
```
[the remark this answers is not in the document]
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/ReassignmentHandler.java:33]: Done.

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/ReassignmentHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/ReassignmentHandler.java:24]: Ah this was asking about future message delivery, will fix before submitting.

[the remark this answers is not in the document]
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/ReassignmentHandler.java:24]: Fixed, we now wait for subscriber shutdown before acknowledging reassignment.

[the remark this answers is not in the document]
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/ReassignmentHandler.java:24]: This now applies to acks as well, added to the comment mentioning them.

**Pull request merged**